### PR TITLE
Methods of Cosmology

### PR DESCRIPTION
**Pull request by nikfilippas**
I have added few lines of code to `core.py` to make (almost) all of the `pyccl` functions which accept `cosmo` as their first argument, methods of the `Cosmology` class.

### New methods, organized by module

1. **pyccl.background**
  - growth_factor
  - growth_factor_unnorm
  - growth_rate
  - comoving_radial_distance
  - comoving_angular_distance
  - angular_diameter_distance
  - luminosity_distance
  - h_over_h0
  - distance_modulus
  - scale_factor_of_chi
  - omega_x
  - rho_x

2. **pyccl.bcm**
  - bcm_model_fka
  - bcm_correct_pk2d

3. **pyccl.boltzmann**
 - get_camb_pk_lin
 - get_isitgr_pk_lin
 - get_class_pk_lin

4. **pyccl.cls**
  - angular_cl

5. **pyccl.correlations**
  - correlation
  - correlation_3d
  - correlation_multipole
  - correlation_3dRsd
  - correlation_3dRsd_avgmu
  - correlation_pi_sigma

6. **pyccl.covariances**
  - angular_cl_cov_cNG

7. **pyccl.pk2d**
  - parse_pk2d

8. **pyccl.power**
  - linear_power
  - nonlin_power
  - linear_matter_power
  - nonlin_matter_power
  - sigmaM
  - sigmaR
  - sigmaV
  - sigma8
  - kNL

9. **pyccl.tracers**
  - get_density_kernel
  - get_lensing_kernel
  - get_kappa_kernel

10. **pyccl.halos.halo_model**
  - halomod_mean_profile_1pt
  - halomod_bias_1pt
  - halomod_power_spectrum
  - halomod_Pk2D
  - halomod_trispectrum_1h
  - halomod_Tk3D_1h

11. **pyccl.halos.massdef**
  - mass2radius_lagrangian
  - convert_concentration   ==> I added this one in `__init__.py` because it's useful

12. **pyccl.nl_pt.power**
  - get_pt_pk2d

13. **pyccl.nl_pt.tracers**
  - translate_IA_norm

### These functions take `cosmo` as their first argument but were not added (because they are not in `__init__.py`)
- _get_concentration
- _get_mf_hb
- _Sig_MG

### These functions take `cosmo` as their first argument but were not added (because they are deprecated)
  - einasto_profile_3d
  - halo_bias
  - halo_concentration
  - halomodel_matter_power
  - hernquist_profile_3d
  - massfunc
  - massfunc_m2r
  - nfw_profile_2d
  - nfw_profile_3d
  - onehalo_matter_power
  - twohalo_matter_power

### Advantages of this method
1. New functions do not have to be manually added in `Cosmology`
2. It is backwards-compatible! (I think)

### CODE REVIEW

**Comment by nikfilippas** [comment]
Does it make sense to have this unit tested? I am not sure because it's kind of tautological.

**Comment by c-d-leonard** [comment]
Re: unit tests - I think just need to have some tests that ensure if we do try to call these functions via method to cosmology we don't get errors. Benchmark tests I would say are not necessary here if it's just changing the piping of identical code to make them methods.

I'm a bit confused about why those functions aren't in __init__.py, I want to check if that's just an oversight for some of them ..

**Comment by nikfilippas** [comment]
I've unit tested this now. The test checks that all functions in pyccl (that are exposed in `__init__`) have been added as attributes of the `Cosmology` object.

Regarding why some of them are not in `__init__`, maybe that was done to discourage direct calls to the Boltzmann solvers as this is handled separately in CCL. I don't mind adding them in this PR if people agree.

**Comment by nikfilippas** [comment]
I think this is ready for review now. Tagging both of you because you were interested in this PR.

**Comment by nikfilippas** [comment]
I also added the boltzmann solvers in `__init__.py` now, so it's easy to get the `Pk2D` objects directly.

File: pyccl/core.py [hunk 1]
```diff
@@ -194,6 +194,23 @@ class Cosmology(object):
...
+            pk2d, power, tracers, halos, nl_pt]
+    funcs = [getmembers(sub, isfunction) for sub in subs]
+    funcs = [func for sub in funcs for func in sub]
+    for name, func in funcs:
+        pars = signature(func).parameters
+        if list(pars)[0] == "cosmo":
+            vars()[name] = func
+    del background, boltzmann, bcm, cls, correlations, covariances, \
+        pk2d, power, tracers, halos, nl_pt
+
```
Comment (damonge) [inline pyccl/core.py:213]: Can we check how much extra time all these new lines add to the creation of a standard cosmology?
I'd like to check this both on a laptop and on a server (@nikfilippas you can use glamdring).

File: pyccl/__init__.py [hunk 2]
```diff
@@ -24,6 +24,9 @@
     h_over_h0, luminosity_distance, distance_modulus, scale_factor_of_chi, \
     omega_x, rho_x
 
+# Boltzmann solvers
+from .boltzmann import get_camb_pk_lin, get_isitgr_pk_lin, get_class_pk_lin
```
Comment (damonge) [inline pyccl/__init__.py:28]: why did you need this?

File: pyccl/core.py [hunk 3]
```diff
@@ -194,6 +194,23 @@ class Cosmology(object):
...
+        cls, correlations, covariances, \
+        pk2d, power, tracers, halos, nl_pt
+    subs = [background, boltzmann, bcm, cls, correlations, covariances,
+            pk2d, power, tracers, halos, nl_pt]
+    funcs = [getmembers(sub, isfunction) for sub in subs]
+    funcs = [func for sub in funcs for func in sub]
+    for name, func in funcs:
+        pars = signature(func).parameters
+        if list(pars)[0] == "cosmo":
+            vars()[name] = func
```
Comment (damonge) [inline pyccl/core.py:210]: can you explain what this line is doing?

File: pyccl/__init__.py (hunk 2 above)
Comment (nikfilippas) [reply pyccl/__init__.py:28]: Why not? It might be useful for anyone who wants to get the camb or class power spectra in a `Pk2D` object in a single line of code. Plus, these functions are not prepended with an underscore, so I don't see why they shouldn't be exposed.

File: pyccl/core.py (hunk 3 above)
Comment (nikfilippas) [reply pyccl/core.py:210]: For every function in the functions exposed in the `__init__.py` files, check the function signature. If the first argument of the funciton is "cosmo", then add it to the local (class) variables as an attribute with the same name.

Comment (nikfilippas) [reply pyccl/core.py:210]: If it's not clear in the source code, I can always add comments! Let me know.

File: pyccl/core.py (hunk 1 above)
Comment (nikfilippas) [reply pyccl/core.py:213]: I've checked it with Spyder's built-in `%timeit` functionality, and, within the standard deviation, there is no difference at all.
That was with a loop of 70,000 instantiations.

Comment (nikfilippas) [reply pyccl/core.py:213]: - with: 42.3 µs ± 135 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
- without: 40.1 µs ± 198 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

Comment (damonge) [reply pyccl/core.py:213]: ok, good

File: pyccl/core.py (hunk 3 above)
Comment (damonge) [reply pyccl/core.py:210]: Yeah, can you add a comment? Took me a bit to figure out what `vars()` (with no argument) did

File: pyccl/core.py (hunk 1 above)
Comment (damonge) [reply pyccl/core.py:213]: @nikfilippas did you test this on glamdring?

File: pyccl/tests/test_cosmology.py [hunk 4]
```diff
@@ -10,6 +10,26 @@
...
+    subs = [background, bcm, cls, correlations, covariances,
+            pk2d, power, tracers, halos, nl_pt]
+    funcs = [getmembers(sub, isfunction) for sub in subs]
+    funcs = [func for sub in funcs for func in sub]
+    for name, func in funcs:
+        pars = signature(func).parameters
+        if list(pars)[0] == "cosmo":
+            _ = getattr(cosmo, name)
+
+
```
Comment (damonge) [inline pyccl/tests/test_cosmology.py:32]: Can you add a test that runs some of the functions (e.g. pick 5 of them with different complexities) as a `Cosmology` method and as a standalone function, and checks that the result is the same?

**Review by damonge — CHANGES_REQUESTED** [review]

File: pyccl/core.py (hunk 1 above)
Comment (nikfilippas) [reply pyccl/core.py:213]: Timings on glamdring:
- with: `37.0 μs per loop (10000 loops)`
- without: `37.3 μs per loop (10000 loops)`

No statistically significant difference then.

File: pyccl/core.py (hunk 3 above)
Comment (nikfilippas) [reply pyccl/core.py:210]: I actually just realised there is already a comment above the loop, explaining what I do in the lines below. Is this enough explanation?

File: pyccl/tests/test_cosmology.py (hunk 4 above)
Comment (nikfilippas) [reply pyccl/tests/test_cosmology.py:46]: Done.

**Review by damonge — APPROVED** [review]
Ok, looks good to me!

File: pyccl/tests/test_cosmology.py [hunk 5]
```diff
@@ -10,6 +10,40 @@
 import pyccl as ccl
 
 
+def test_cosmo_methods():
+    """ Check that all pyccl functions that take cosmo
+    as their first argument are methods of the Cosmology object.
+    """
+    from inspect import getmembers, isfunction, signature
+    from pyccl import background, bcm, \
```
Comment (nikfilippas) [inline pyccl/tests/test_cosmology.py:18]: I am not testing for all the modules here. Commit incoming...

Comment (nikfilippas) [reply pyccl/tests/test_cosmology.py]: Done. I am now testing the `boltzmann` module as well.
Other than that, I added `neutrinos` and `tk3d` even though they contain no functions with `cosmo`, just in case there are any future additions in there (so that this part of the code won't need maintenance).
@damonge do you still approve?

Comment (damonge) [reply pyccl/tests/test_cosmology.py]: looks good to me, yes.

**Review by damonge — APPROVED** [review]
